File: test/virtual_inheritance.cpp
// Copyright 2009 Troy D. Straszheim
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
//
// Big thanks to Roman Yakovenko for this test case
//

#include <boost/python.hpp>
#include <boost/python/extract.hpp>

struct base
{
  virtual std::string f() { return "base::f()"; }
  virtual std::string g() { return "base::g()"; }
};
	
struct derived : virtual base
{ 
  virtual std::string f() { return "derived::f()"; }
  virtual std::string g() { return "derived::g()"; }
};    

namespace bp = boost::python;

struct base_wrapper : base, bp::wrapper<base> 
{
  base_wrapper(base const& arg)
    : base(arg), bp::wrapper<base>()
  { }

  base_wrapper()
    : base(), bp::wrapper<base>()
  { }

  //
  //  Here's one way to do it
  //
  virtual std::string f() 
  {
    if(bp::override func_f = this->get_override("f"))
      return func_f();
    else
      return this->base::f();
  }

  //
  //  Here's another
  //
  virtual std::string g()
  {
    if(bp::override func_g = this->get_override("g"))
      return func_g();
    else
      return base::g();
  }
  std::string default_g()
  {
    return base::g();
  }
};

struct derived_wrapper 
  : derived, bp::wrapper<derived> 
{
  derived_wrapper(derived const& arg) 
    : derived(arg), bp::wrapper<derived>()
  { }

  derived_wrapper()
    : derived(), bp::wrapper<derived>()
  { }

  virtual std::string f() 
  {
    if( bp::override func_f = this->get_override("f"))
      return func_f();
    else
      return this->derived::f();
  }

  virtual std::string g()
  {
    if(bp::override func_g = this->get_override("g"))
      return func_g();

    return derived::g();
  }
  std::string default_g()
  {
    return std::string("derived_wrapper::default_g() ") + derived::g();
  }

};

std::string call_base_f(base& b) { return b.f(); }
std::string call_base_g(base& b) { return b.g(); }

std::string call_derived_f(derived& d) { return d.f(); }
std::string call_derived_g(derived& d) { return d.g(); }

BOOST_PYTHON_MODULE(virtual_inheritance_ext)
{
  bp::class_<base_wrapper>("base")    
    .def("f", &base_wrapper::f)
    .def("g", &base::g, &base_wrapper::g)
    ;

  bp::class_<derived_wrapper, bp::bases<base> >("derived")    
    .def("f", &derived_wrapper::f)
    .def("g", &base::g, &derived_wrapper::default_g)
    ;

  bp::def("call_base_f", call_base_f);
  bp::def("call_base_g", call_base_g);

  bp::def("call_derived_f", call_derived_f);
  bp::def("call_derived_g", call_derived_g);
}

File: test/virtual_inheritance.py
#!/usr/bin/env python
#
#
# Copyright Troy D. Straszheim 2009. Distributed under the Boost
# Software License, Version 1.0. (See accompanying file
# LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
#
#
# Check that virtual inheritance works
#
"""
>>> from virtual_inheritance_ext import *

    Base is just base.

>>> b = base()
>>> b.f()
'base::f()'
>>> b.g()
'base::g()'

    Derived is just derived.
 
>>> d = derived()
>>> d.f()
'derived::f()'
>>> d.g()
'derived_wrapper::default_g() derived::g()'

    Inherit from base, don't redefine f()

>>> class B1(base):
...     def __init__(self):
...         base.__init__(self)
... 
>>> b = B1()
>>> b.f()
'base::f()'
>>> call_base_f(b)
'base::f()'
>>> call_base_g(b)
'base::g()'

    Inherit from base and implement f() only
   
>>> class B2(base):
...     def __init__(self):
...         base.__init__(self)
...     def f(self):
...         return 'B2.f'
... 
>>> b = B2()
>>> b.f()
'B2.f'
>>> call_base_f(b)
'B2.f'
>>> call_base_g(b)
'base::g()'

    Inherit from derived, don't override f() 

>>> class D1(derived):
...     def __init__(self):
...         derived.__init__(self)
... 
>>> d1 = D1()
>>> d1.f()
'derived::f()'
>>> call_derived_f(d1)
'derived::f()'
>>> call_base_f(d1)
'derived::f()'

    Inherit from derived, do override f()

>>> class D2(derived):
...     def __init__(self):
...         derived.__init__(self)
...     def f(self):
...         return 'D2.f'
... 
>>> d2 = D2()
>>> d2.f()
'D2.f'
>>> call_base_f(d2)
'D2.f'
>>> call_derived_f(d2)
'D2.f'
>>>
"""
